src/shims/dlsym.rs

```diff
@@ -18,7 +18,8 @@ impl Dlsym {
     pub fn from_str(name: &[u8], target_os: &str) -> InterpResult<'static, Option<Dlsym>> {
         let name = &*String::from_utf8_lossy(name);
         Ok(match target_os {
-            "linux" | "macos" => posix::Dlsym::from_str(name, target_os)?.map(Dlsym::Posix),
+            "linux" | "macos" | "android" =>
```

Comment:
Looking at this, I think the new helper function should probably return a `bool` (rather than assert) and be used here as well.

Reply:
How would you style this? What immediately popped in my head looks awkward. I think there's value in retaining the match for future custom targets (redox or something?) if desired.

```patch
diff --git a/src/helpers.rs b/src/helpers.rs
index 2e870004..1e38e074 100644
--- a/src/helpers.rs
+++ b/src/helpers.rs
@@ -442,15 +442,22 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     /// if this is not the case.
     fn assert_target_os_is_unix(&self, name: &str) {
         assert!(
-            matches!(
-                self.eval_context_ref().tcx.sess.target.os.as_str(),
-                "linux" | "macos" | "android"
-            ),
+            self.target_os_is_unix(),
             "`{}` is only available for the UNIX target family",
             name,
         );
     }
 
+    /// Helper function used inside the shims of foreign functions to assert that the target OS
+    /// is `target_os`. It panics showing a message with the `name` of the foreign function
+    /// if this is not the case.
+    fn target_os_is_unix(&self) -> bool {
+        matches!(
+            self.eval_context_ref().tcx.sess.target.os.as_str(),
+            "linux" | "macos" | "android"
+        )
+    }
+
     /// Get last error variable as a place, lazily allocating thread-local storage for it if
     /// necessary.
     fn last_error_place(&mut self) -> InterpResult<'tcx, MPlaceTy<'tcx, Tag>> {
diff --git a/src/shims/foreign_items.rs b/src/shims/foreign_items.rs
index 64ec9dcf..f1a1db1c 100644
--- a/src/shims/foreign_items.rs
+++ b/src/shims/foreign_items.rs
@@ -687,8 +687,8 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
             // Platform-specific shims
             _ => match this.tcx.sess.target.os.as_str() {
-                "linux" | "macos" | "android" => return shims::posix::foreign_items::EvalContextExt::emulate_foreign_item_by_name(this, link_name, abi, args, dest, ret),
                 "windows" => return shims::windows::foreign_items::EvalContextExt::emulate_foreign_item_by_name(this, link_name, abi, args, dest, ret),
+                _ if this.target_os_is_unix() => return shims::posix::foreign_items::EvalContextExt::emulate_foreign_item_by_name(this, link_name, abi, args, dest, ret),
                 target => throw_unsup_format!("the target `{}` is not supported", target),
             }
         };
```

Reply:
I wouldn't swap the order in the `match`; seems fine to me otherwise.

You could also consider making `target_os_is_unix` take the string and then writing
```rust
  target if target_os_is_unix(target) =>
```
which is a bit cleaner since it still actually tests the value being matched.